components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigration.kt

```diff
@@ -0,0 +1,216 @@
+        if (migrationsToRun.isEmpty()) {
+            logger.debug("No migrations to run.")
+            val result = CompletableDeferred<MigrationsResult>()
+            result.complete(emptyMap())
+            return result
+        }
+
+        return runMigrationsAsync(migrationsToRun)
+    }
+
+    private fun runMigrationsAsync(
+        migrations: List<Pair<Migration, Int>>
+    ): Deferred<MigrationsResult> = CoroutineScope(coroutineContext).async { synchronized(migrationLock) {
+        val results = mutableMapOf<Migration, Pair<Int, MigrationResult>>()
+
+        migrations.forEach { versionedMigration ->
+            logger.debug("Migrating $versionedMigration")
+
+            val migrationResult = when (versionedMigration.first) {
+                Migration.HISTORY -> migrateHistory()
+                Migration.BOOKMARKS -> migrateBookmarks()
+                Migration.OPEN_TABS -> migrateOpenTabs()
+            }
+
+            results[versionedMigration.first] = versionedMigration.second to migrationResult
+
+            when (migrationResult) {
+                is MigrationResult.Failure -> {
+                    logger.error(
+                        "Failed to migrate $versionedMigration",
+                        migrationResult.throwables.first()
+                    )
+                }
+                is MigrationResult.Success -> {
+                    logger.debug(
+                        "Migrated $versionedMigration"
+                    )
+                }
+                is MigrationResult.SuccessWithData<*> -> {
+                    logger.debug(
+                        "Migrated $versionedMigration with resulting data"
+                    )
+                }
+            }
+        }
+
+        val migrationStore = MigrationHistoryStore(context)
+        migrationStore.setToCache(results.toMigrationHistoryRecord())
+        results
+    } }
+
+    private fun migrateHistory(): MigrationResult {
+        if (dbPath == null) {
+            return MigrationResult.Failure(IllegalStateException("Missing DB path"))
+        }
+        return try {
+            historyStorage.importFromFennec(dbPath)
+            MigrationResult.Success
+        } catch (e: Exception) {
+            MigrationResult.Failure(e)
+        }
+    }
+
+    private fun migrateBookmarks(): MigrationResult {
+        if (dbPath == null) {
+            return MigrationResult.Failure(IllegalStateException("Missing DB path"))
+        }
+        return try {
+            bookmarksStorage.importFromFennec(dbPath)
+            MigrationResult.Success
+        } catch (e: Exception) {
+            MigrationResult.Failure(e)
+        }
+    }
+
+    private fun migrateOpenTabs(): MigrationResult {
+        if (profile == null) {
+            return MigrationResult.Failure(IllegalStateException("Missing Profile path"))
+        }
+        return try {
+            FennecSessionMigration.migrate(File(profile.path))
```

Comment:
What's your plan for the result here? I would like the migration code to take care of taking the result and injecting this into the sessionmanager/store.

Reply:
As-is, that logic would need to live in Fenix. If we can configure `FennecMigration` with the sessionStore, then `migrateOpenTabs` can do the injection.